frontend/src/metabase/common/hooks/use-stacked-modals/useStackedModals.ts

```diff
@@ -0,0 +1,81 @@
+import { useState } from "react";
+import { useMount } from "react-use";
+import _ from "underscore";
+
+type Props<TKeys extends string> = {
+  modals: TKeys[];
+  defaultOpened?: TKeys;
+  withOverlay: boolean;
+};
+
+type GetModalPropsReturn = {
+  /**
+   * to support both Sidesheet and mantine's Modal property.
+   * To fix it rename Sidesheet's prop
+   */
+  opened: boolean;
+  isOpen: boolean;
+  onClose: () => void;
+  closeOnEscape: boolean;
+  withOverlay: boolean;
+  withTransparentOverlay: boolean;
+};
+
+export const useStackedModals = <TKeys extends string>({
+  modals,
+  defaultOpened,
+  withOverlay = true,
+}: Props<TKeys>) => {
+  const [state, setState] = useState<Partial<Record<TKeys, boolean>>>(() =>
+    modals.reduce((memo: Partial<Record<TKeys, boolean>>, key) => {
+      memo[key] = false;
+      return memo;
+    }, {}),
+  );
+  const [stack, setStack] = useState<TKeys[]>([]);
+
+  const currentModal = stack.at(-1);
+  const firstOpenedModal = stack.at(0);
+
+  const close = (key: TKeys) => {
+    setState((prev) => ({ ...prev, [key]: false }));
+    setStack((prev) => prev.slice(0, -1));
+  };
+
+  const open = (key: TKeys) => {
+    setState((prev) => ({ ...prev, [key]: true }));
+    setStack((prev) => [...prev, key]);
+  };
+
+  useMount(() => {
+    if (!defaultOpened) {
+      return;
+    }
+    // the modal is not rendered until it is "open"
+    // but we want to set it open after it mounts to get
+    // pretty animations
+    open(defaultOpened);
+  });
+
+  const getModalProps = (
+    key: TKeys,
+    // TODO(egorgrushin): introduce mergeProps pattern
+    override?: Partial<GetModalPropsReturn>,
+  ): GetModalPropsReturn =>
+    _.defaults(override, {
+      isOpen: state[key] ?? false,
+      opened: state[key] ?? false,
+      onClose: () => close(key),
+      closeOnEscape: key === currentModal,
+      withOverlay,
+      /** Invisible overlay to prevent double darkening while preserving click-outside handling */
+      withTransparentOverlay: withOverlay && key !== firstOpenedModal,
+    });
```

Comment:
:x: This is too magic.

We've had our share of pain dealing with props spreading in this project.
It makes it difficult to read code and track the flow of information.

Explicit props are almost always better (there [are exceptions](https://github.com/metabase/metabase/blob/f12e3320190ef52a325fefbab9e3a5a0ea5324ba/frontend/src/metabase/ui/components/feedback/Loader/Loader.tsx#L12-L14), of course).

Reply:
This is very handy way to make powerful things in react - it's called Prop Getters pattern (or something like this). Can we discuss it more before we deny it? 
If it's typed correctly, it's easy to track the flow of information.

Reply:
```tsx
{...getModalProps("caching", { onClose: handleClose })}
```

The only way to figure out what props are actually passed to the component is to read the `getModalProps` definition, and then come back to the consumer and see if any of these props were later overridden.

This makes it better:

```tsx
const { closeOnEscape, isOpen, withOverlay, onClose  } = getModalProps("caching", { onClose: handleClose });

/* ... */

closeOnEscape={closeOnEscape}
isOpen={isOpen}
withOverlay={withOverlay}
onClose={onClose}
```

And with suggestion from lines 62-63, even better:

```tsx
const { closeOnEscape, isOpen, withOverlay  } = getModalProps("caching");

/* ... */

closeOnEscape={closeOnEscape}
isOpen={isOpen}
withOverlay={withOverlay}
onClose={handleClose}
```